src/engine/sparqlExpressions/StringExpressions.cpp

```diff
@@ -467,12 +470,54 @@ class ConcatExpression : public detail::VariadicExpression {
     // We store the (intermediate) result either as single string or a vector.
     // If the result is a string, then all the previously evaluated children
     // were constants (see above).
+
     std::variant<std::string, StringVec> result{std::string{""}};
+    StringVec dataTypesOrLangTags(ctx->_allocator);
+
+    auto storeDatatypeOrLanguageTag = [&](const auto& literal) {
+      if (literal.hasDatatype()) {
+        dataTypesOrLangTags.push_back(
+            std::string{asStringViewUnsafe(literal.getDatatype())});
+      } else if (literal.hasLanguageTag()) {
+        dataTypesOrLangTags.push_back(
+            std::string{asStringViewUnsafe(literal.getLanguageTag())});
+      } else {
+        dataTypesOrLangTags.push_back("");
+      }
+    };
```

Comment:
This doesn't quite work.
You currently store the datatypes and language tags in a flat vector for all columns AND rows.
You have to do this per row, because for example consider

```
?x -> "en"@en und "fr"@fr
then 
CONCAT(?x, ?x)
should yield
"enen"@en und "frfr"@fr, as the language tags are equal in eacht row.
```
You could for each row store a separate vector of language tag or datatype, but i would only store
one entry which either stores the matching suffix so far, or something empty (because no suffix was found yet), or the information that there already was a mismatch and the suffix will be empty.